### ISSUE TITLE
[BUG] TextField didBeginEditing and didEndEditing are not called correctly

### ISSUE THREAD

**Issue by sgade** [stripe/stripe-ios#2242]
## Summary

`STPPaymentCardTextFieldDelegate`'s `paymentCardTextFieldDidBeginEditing(_:)` and `paymentCardTextFieldDidEndEditing(_:)` functions are not being called correctly. In my observation, the textfield only reports beginning editing once. Further calls do not happen.

## Code to reproduce

The observations were done in a SwiftUI `UIViewRepresentable` instance with a basic coordinator. Create a blank SwiftUI project and use the following implementation:

```swift
struct ContentView: View {

    var body: some View {
        ScrollView {
            StripePaymentCardTextField()
        }
        .scrollDismissesKeyboard(.interactively)
    }

}
```

```swift
import Stripe
import SwiftUI

struct StripePaymentCardTextField: UIViewRepresentable {

    func makeUIView(context: Context) -> STPPaymentCardTextField {
        let textField = STPPaymentCardTextField()
        textField.delegate = context.coordinator
        return textField
    }

    func updateUIView(_ paymentCardField: STPPaymentCardTextField, context: Context) {}

    func makeCoordinator() -> Coordinator {
        Coordinator(parent: self)
    }

    final class Coordinator: NSObject {

        private let parent: StripePaymentCardTextField

        init(parent: StripePaymentCardTextField) {
            self.parent = parent
        }
    }

}

// MARK: - STPPaymentCardTextFieldDelegate

extension StripePaymentCardTextField.Coordinator: STPPaymentCardTextFieldDelegate {

    func paymentCardTextFieldDidBeginEditing(_ textField: STPPaymentCardTextField) {
        print("Begin editing")
    }

    func paymentCardTextFieldDidEndEditing(_ textField: STPPaymentCardTextField) {
        print("End editing")
    }

}
```

## iOS version

* iOS 16.2
* Observed since iOS 15 by my colleagues (apparently not present in iOS 14 and below)

## Installation method

Swift Package Manager

## SDK version

* Current: 23.3.2
* Observed since 21.8.1 (using iOS 15)

## Other information

Upon closer inspection, it seems to me like `getAndUpdateSubviewEditingTransitionState(fromCall:)` called during UITextField delegate functions in `STPPaymentCardTextField.swift` is handling the unmatched calls checks incorrectly. `isMidSubviewEditingTransition` is always true, even if the last text fields resigns first responder and no other text field gains focus next.

**Comment by yuki-stripe** [comment]
Hello @sgade,

Thanks for reporting this and providing the sample code. I've reproduced the problem in SwiftUI, although interestingly I can't reproduce w/ our UIKit-based example code (in our "UI Examples" example project). 

I've added it to our internal tracker but can't promise a timeline for a fix - we are prioritizing improving [our PaymentSheet integration](https://stripe.com/docs/payments/accept-a-payment). Is it possible for you to use that integration instead?
